src/libsyntax/parse/parser.rs

```diff
@@ -6132,6 +6132,22 @@ impl<'a> Parser<'a> {
                 err.span_suggestion_short_with_applicability(
                     self.span, msg, "".to_string(), Applicability::MachineApplicable
                 );
+                if !items.is_empty() {  // Issue #51603
+                    let previous_item = &items[items.len()-1];
+                    let previous_item_kind_name = match previous_item.node {
+                        // say "braced struct" because tuple-structs and
+                        // braceless-empty-struct declarations do take a semicolon
+                        ItemKind::Struct(..) => Some("braced struct"),
+                        ItemKind::Enum(..) => Some("enum"),
+                        ItemKind::Trait(..) => Some("trait"),
+                        ItemKind::Union(..) => Some("union"),
+                        _ => None,
+                    };
+                    if let Some(name) = previous_item_kind_name {
+                        err.help(&format!("{} declarations are not followed by a semicolon",
```

Comment:
This is of course very opinionated, but what do you think about replacing the `expected item, found ;` message with this message?

Reply:
I'm negatively disposed. We're trying to make inferences about the user's state of mind and helpfully remind them how the language works ([in contrast to other languages](https://github.com/rust-lang/rust/issues/46186#issuecomment-346540452) that we think they might be confusing us with). It's a good magical thing to do, but I think it would be overstepping the mandate of the parser to make it the top-line message rather than an ancillary note.

Reply:
I think it's a question of numbers. If most ppl add semicolons because they think the definitions need one, then the error should reflect that. Sadly we don't have numbers...

Reply:
And the issue was about changing the error message ;)

Reply:
@oli-obk 

> I think it's a question of numbers.

While I'm happy to defer to the authority of you or other team members, I don't think numbers are very relevant to the distinction I was trying to draw in my previous comment.

For a perhaps clearer-cut case, consider [what we emit](https://github.com/rust-lang/rust/blob/9fd3d7899a444fe110ae9277542b306844ecf0c4/src/test/ui/did_you_mean/issue-48492-tuple-destructure-missing-parens.stderr) when someone tries to destructure a tuple using comma-separated values and no parentheses (_e.g._, `let a, b = (1, 2);`). In that case, I think it's appropriate that the top-line error is "unexpected `,` in pattern", and only the ancillary suggestion suggests adding parentheses: the former message says what went wrong with parsing (there's nothing in the grammar that lets you put a comma there) while being agnostic about what could have motivated someone to write that code, and _then_, that having been established, we go on to provide a helpful suggestion that maybe you meant to use parentheses. Even if (as seems likely), among people who encounter this error, a solid majority of them geniunely wanted a tuple-unpacking and didn't think parens were necessary, and only a tiny minority just happened to have their fingers slip and insert a comma for no reason, it still makes sense to have both messages in that order, because they're serving different functions: flagging the parsing error is distinct from pointing out what misunderstanding of the language might have motivated that error, and the former should come first because it's an analytic truth, whereas the latter is fundamentally only a guess, though a very high-probability guess it may be. I think the same phenomenon is going on here.

Reply:
Well we definitely err on the safe side by keeping the "expected,found" error. You're right that it would be a major change to diagnostics and should not be done lightly